kubernetes/apps_kubeclient.go

```diff
@@ -770,63 +867,433 @@ func (kc *kubeClient) getPods(namespace string, uid types.UID) ([]v1.Pod, error)
+	return false, false
+}
+
+func isPullingImage(pod *v1.Pod) bool {
+	const containerCreating string = "ContainerCreating"
+	// If pod is pending with a container waiting due to a "ContainerCreating" event,
+	// categorize as "Pulling". This may change as more information is made available.
+	// See: https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js#L663
+	if pod.Status.Phase == v1.PodPending {
+		for _, status := range pod.Status.ContainerStatuses {
+			waiting := status.State.Waiting
+			if waiting != nil && waiting.Reason == containerCreating {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isPodReady(pod *v1.Pod) bool {
+	// If all of the pod's containers have a ready status, then the pod is
+	// considered ready.
+	total := len(pod.Spec.Containers)
+	numReady := 0
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Ready {
+			numReady++
+		}
+	}
+	return numReady == total
+}
+
+func (kc *kubeClient) getBestRoute(namespace string, dc *deployment) (*url.URL, error) {
+	serviceMap, err := kc.getMatchingServices(namespace, dc)
+	if err != nil {
+		return nil, err
+	}
+	// Get routes and associate to services using spec.to.name
+	err = kc.getRoutesByService(namespace, serviceMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find route with highest score according to heuristics from web-console
+	var bestRoute *route
+	bestScore := -1
+	for _, routes := range serviceMap {
+		for _, route := range routes {
+			score := scoreRoute(route)
+			if score > bestScore {
+				bestScore = score
+				bestRoute = route
+			}
+		}
+	}
+
+	// Construct URL from best route
+	var result *url.URL
+	if bestRoute != nil {
+		scheme := "http"
+		if bestRoute.tls {
+			scheme = "https"
+		}
+		result = &url.URL{
+			Scheme: scheme,
+			Host:   bestRoute.host,
+			Path:   bestRoute.path,
+		}
 	}
 
 	return result, nil
 }
 
+func (kc *kubeClient) getMatchingServices(namespace string, dc *deployment) (routesByService map[string][]*route, err error) {
+	services, err := kc.Services(namespace).List(metaV1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// Check if each service's selector matches labels in deployment's pod template
+	template := dc.current.Spec.Template
+	if template == nil {
+		return nil, errors.New("No pod template for current deployment")
+	}
+	routesByService = make(map[string][]*route)
+	for _, service := range services.Items {
+		selector := service.Spec.Selector
+		match := true
+		// Treat empty selector as not matching
+		if len(selector) == 0 {
+			match = false
+		}
+		for key := range selector {
+			if selector[key] != template.Labels[key] {
+				match = false
+			}
+		}
+		// If all selector labels match those in the pod template, add service key to map
+		if match {
+			routesByService[service.Name] = nil
+		}
+	}
+	return routesByService, nil
+}
+
+func (kc *kubeClient) getRoutesByService(namespace string, routesByService map[string][]*route) error {
```

Comment:
is this method covered by 1 or more tests ?

Reply:
A whitebox test has not yet been written.